ui/src/dashboards/components/QueryEditor.tsx

```diff
@@ -0,0 +1,465 @@
+                  )}
+                  onClick={this.handleUpdate}
+                >
+                  Submit Query
+                </button>
+              </QueryStatus>
+            </div>
+            <div className="varmoji-back">
+              {isTemplating ? (
+                <TemplateDrawer
+                  onClickTempVar={this.handleClickTempVar}
+                  templates={filteredTemplates}
+                  selected={selectedTemplate}
+                  onMouseOverTempVar={this.handleMouseOverTempVar}
+                  handleClickOutside={this.handleCloseDrawer}
+                />
+              ) : null}
+            </div>
+          </div>
+        </div>
+      </div>
+    )
+  }
+
+  private hideTemplateValues = (): void => {
+    this.setQueryValue(this.state.queryText)
+  }
+
+  private handleToggleFocus = (e: MouseEvent<HTMLButtonElement>): void => {
+    e.stopPropagation()
+    e.preventDefault()
+
+    this.editor.focus()
+  }
+
+  private handleToggleTemplateValues = (): void => {
+    const {isShowingTemplateValues} = this.state
+
+    if (isShowingTemplateValues) {
+      this.hideTemplateValues()
+    } else {
+      this.showTemplateValues()
+    }
+  }
+
+  private get isDisabled(): boolean {
+    return this.state.isTemplating || this.state.isShowingTemplateValues
+  }
+
+  private get queryCodeClassName(): string {
+    const {focused} = this.state
+
+    return classnames('query-editor--code', {focus: focused})
+  }
+
+  private get templateToggleStatus(): string {
+    if (this.state.isShowingTemplateValues) {
+      return HIDE_TEMPLATES
+    }
+
+    return SHOW_TEMPLATES
+  }
+
+  private showTemplateValues = async (): Promise<void> => {
+    const {queryText} = this.state
+
+    const queryWithTemplateValues = await replaceQueryTemplates(
+      queryText,
+      this.props.config.source,
+      this.props.templates,
+      NULL_RESOLUTION
+    )
+
+    this.setState({
+      editorValue: queryWithTemplateValues,
+      isShowingTemplateValues: true,
+    })
+
+    if (this.editor) {
+      this.editor.focus()
+    }
+  }
+
+  private handleCloseDrawer = (): void => {
+    this.setState({isTemplating: false})
+  }
+
+  private handleMouseOverTempVar = (template: Template) => () => {
+    this.handleTemplateReplace(template, false)
+  }
+
+  private handleClickTempVar = (template: Template) => () => {
+    // Clicking a tempVar does the same thing as hitting 'Enter'
+    this.handleTemplateReplace(template, true)
+    this.closeDrawer()
+  }
+
+  private handleMountEditor = (editor: IInstance) => {
+    this.editor = editor
+  }
+
+  private handleUnmountEditor = () => {
+    this.editor = null
+  }
+
+  private handleBlurEditor = (): void => {
+    this.setState({focused: false})
+
+    this.hideTemplateValues()
+  }
+
+  private handleFocusEditor = (): void => {
+    this.setState({focused: true})
+  }
+
+  private handleKeyDownEditor = (__, e) => {
```

Comment:
would you consider adding the return type here and on other methods missing them? for dev readability/expectation purposes, mostly